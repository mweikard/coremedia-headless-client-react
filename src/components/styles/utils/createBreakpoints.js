// @flow
import { em } from 'polished';

import type { Breakpoint, BreakpointsInput, BreakpointsOutput } from '../../../types';

const keys: Array<Breakpoint> = ['xs', 'sm', 'md', 'lg', 'xl'];

const createBreakpoints = (breakpoints: BreakpointsInput): BreakpointsOutput => {
  const {
    values = {
      xs: 0,
      sm: 768,
      md: 992,
      lg: 1280,
      xl: 1920,
    },
    step = 5,
  } = breakpoints;

  const min = (key: Breakpoint | number) => {
    const value =
      typeof key === 'string' && typeof values[key] === 'number' ? em(values[key]) : `${key}px`;
    return `@media (min-width:${value})`;
  };

  const max = (key: Breakpoint | number) => {
    let value;

    if (typeof key === 'number') {
      value = em(key);
    } else {
      const endIndex = keys.indexOf(key) + 1;

      if (endIndex === keys.length) {
        return min('xs');
      }

      const upperLimit = values[keys[endIndex]];

      value =
        typeof upperLimit === 'number' && endIndex > 0 ? em(upperLimit - step / 100) : `${key}px`;
    }
    return `@media (max-width:${value})`;
  };

  const between = (start: Breakpoint | number, end: Breakpoint | number) => {
    const minValue =
      typeof start === 'string' && typeof values[start] === 'number'
        ? em(values[start])
        : `${start}px`;

    let maxValue;

    if (typeof end === 'number') {
      maxValue = em(end);
    } else {
      const endIndex = keys.indexOf(end) + 1;

      if (endIndex === keys.length) {
        return min(start);
      }

      const upperLimit = values[keys[endIndex]];

      maxValue =
        typeof upperLimit === 'number' && endIndex > 0 ? em(upperLimit - step / 100) : `${end}px`;
    }
    return `@media (min-width:${minValue}) and (max-width:${maxValue})`;
  };

  const exact = (key: Breakpoint) => between(key, key);

  return {
    keys,
    values,
    min,
    max,
    between,
    exact,
  };
};

export default createBreakpoints;
