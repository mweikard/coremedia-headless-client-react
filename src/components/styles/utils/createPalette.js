// @flow
import { rgb, rgba } from 'polished';
import merge from 'deepmerge';

import type { PaletteInput, PaletteOutput } from '../../../types';

const createPalette = (palette: PaletteInput): PaletteOutput => {
  const {
    primary = {
      color: rgb(10, 132, 255),
      text: rgba(255, 255, 255, 0.87),
    },
    secondary = {
      color: rgb(237, 0, 181),
      text: rgba(255, 255, 255, 0.87),
    },
    background = {
      color: rgb(255, 255, 255),
      textPrimary: rgba(0, 0, 0, 0.87),
      textSecondary: rgba(0, 0, 0, 0.6),
      textDisabled: rgba(0, 0, 0, 0.38),
    },
    surface = {
      color: rgb(255, 255, 255),
      textPrimary: rgba(0, 0, 0, 0.87),
      textSecondary: rgba(0, 0, 0, 0.6),
      textDisabled: rgba(0, 0, 0, 0.38),
    },
    overlay = {
      color: rgba(173, 173, 173, 0.9),
    },
    ...rest
  } = palette;

  const common = {
    white: rgb(255, 255, 255),
    black: rgb(0, 0, 0),
  };

  return merge(
    {
      primary,
      secondary,
      background,
      surface,
      overlay,
      common,
      additional: {},
    },
    rest
  );
};

export default createPalette;
