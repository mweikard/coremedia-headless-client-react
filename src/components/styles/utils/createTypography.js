// @flow
import { rem, em } from 'polished';
import merge from 'deepmerge';

import type {
  PaletteOutput,
  BreakpointsOutput,
  TypographyInput,
  TypographyOutput,
} from '../../../types';

const createTypography = (
  palette: PaletteOutput,
  breakpoints: BreakpointsOutput,
  typography: TypographyInput
): TypographyOutput => {
  const {
    fontFamily = 'Helvetica, Arial, sans-serif',
    fontWeightLight = 300,
    fontWeightRegular = 400,
    fontWeightMedium = 500,
    htmlFontSize = 16,
    ...rest
  } =
    typeof typography === 'function' ? typography(palette, breakpoints) : typography;

  return merge(
    {
      fontFamily,
      fontWeightLight,
      fontWeightRegular,
      fontWeightMedium,
      htmlFontSize,
      title1: {
        fontFamily: fontFamily,
        fontSize: rem(34, htmlFontSize),
        fontWeight: fontWeightLight,
        textTransform: 'uppercase',
        lineHeight: em(43, 34),
        color: palette.background.textSecondary,
      },
      title2: {
        fontFamily: fontFamily,
        fontSize: rem(30, htmlFontSize),
        fontWeight: fontWeightLight,
        textTransform: 'uppercase',
        lineHeight: em(34, 30),
        color: palette.background.textSecondary,
      },
      title3: {
        fontFamily: fontFamily,
        fontSize: rem(26, htmlFontSize),
        fontWeight: fontWeightLight,
        textTransform: 'uppercase',
        lineHeight: em(30, 26),
        color: palette.background.textSecondary,
      },
      title4: {
        fontFamily: fontFamily,
        fontSize: rem(23, htmlFontSize),
        fontWeight: fontWeightRegular,
        textTransform: 'uppercase',
        lineHeight: em(29, 23),
        color: palette.background.textSecondary,
      },
      title5: {
        fontFamily: fontFamily,
        fontSize: rem(20, htmlFontSize),
        fontWeight: fontWeightRegular,
        lineHeight: em(26, 20),
        color: palette.background.textPrimary,
      },
      title6: {
        fontFamily: fontFamily,
        fontSize: rem(18, htmlFontSize),
        fontWeight: fontWeightRegular,
        lineHeight: em(23, 18),
        color: palette.background.textPrimary,
      },
      body1: {
        fontFamily: fontFamily,
        fontSize: rem(16, htmlFontSize),
        fontWeight: fontWeightRegular,
        lineHeight: em(21, 16),
        color: palette.background.textPrimary,
      },
      body2: {
        fontFamily: fontFamily,
        fontSize: rem(14, htmlFontSize),
        fontWeight: fontWeightRegular,
        lineHeight: em(18, 14),
        color: palette.background.textPrimary,
      },
      button: {
        fontFamily: fontFamily,
        fontSize: rem(16, htmlFontSize),
        fontWeight: fontWeightMedium,
        textTransform: 'none',
        color: palette.background.textPrimary,
      },
      caption: {
        fontFamily: fontFamily,
        fontSize: rem(12, htmlFontSize),
        fontWeight: fontWeightRegular,
        textTransform: 'none',
        lineHeight: em(16, 12),
        color: palette.background.textSecondary,
      },
    },
    rest
  );
};

export default createTypography;
