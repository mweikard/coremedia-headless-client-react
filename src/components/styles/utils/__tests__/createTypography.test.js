// @flow
import createTypography from '../createTypography';
import createPalette from '../createPalette';
import createBreakpoints from '../createBreakpoints';

describe('createTypography()', () => {
  const palette = createPalette({});
  const breakpoints = createBreakpoints({});
  it(`should return matching value using default parameter value`, () => {
    const result = createTypography(palette, breakpoints, {});
    expect(result).toMatchSnapshot();
  });
  it(`should return matching value using custom value (object)`, () => {
    const typography = {
      fontFamily: 'Verdana',
      fontWeightLight: 200,
      fontWeightRegular: 300,
      fontWeightMedium: 700,
      htmlFontSize: 14,
    };
    const result = createTypography(palette, breakpoints, typography);
    expect(result).toMatchSnapshot();
  });
  it(`should return matching value using custom value (function)`, () => {
    const fontFamilyHeading = '"Comic Sans"';
    const fontFamilyBase = 'Georgia';
    const fontWeightLight = 200;
    const fontWeightRegular = 300;
    const fontWeightMedium = 700;
    const htmlFontSize = 14;
    const typography = (palette, breakpoints) => ({
      fontFamily: fontFamilyBase,
      fontWeightLight,
      fontWeightRegular,
      fontWeightMedium,
      htmlFontSize,
      title1: {
        fontFamily: fontFamilyHeading,
        fontSize: '22px',
        fontWeight: fontWeightLight,
        textTransform: 'uppercase',
        lineHeight: '1.6em',
        [breakpoints.min('sm')]: {
          fontSize: '34px',
        },
      },
      title2: {
        fontFamily: fontFamilyHeading,
        fontSize: '22px',
        fontWeight: fontWeightLight,
        textTransform: 'uppercase',
        lineHeight: '1.5em',
        [breakpoints.min('sm')]: {
          fontSize: '30px',
        },
      },
      title3: {
        fontFamily: fontFamilyHeading,
        fontSize: '22px',
        fontWeight: fontWeightLight,
        textTransform: 'uppercase',
        lineHeight: '1.4em',
        [breakpoints.min('sm')]: {
          fontSize: '26px',
        },
      },
      title4: {
        fontFamily: fontFamilyHeading,
        fontSize: '22px',
        fontWeight: fontWeightRegular,
        textTransform: 'uppercase',
        lineHeight: '1.3em',
        [breakpoints.min('sm')]: {
          fontSize: '23px',
        },
      },
      title5: {
        fontFamily: fontFamilyBase,
        fontSize: '20px',
        fontWeight: fontWeightRegular,
        lineHeight: '1.2em',
      },
      title6: {
        fontFamily: fontFamilyBase,
        fontSize: '16px',
        fontWeight: fontWeightRegular,
        lineHeight: '1.1em',
        [breakpoints.min('sm')]: {
          fontSize: '18px',
        },
      },
      body1: {
        fontFamily: fontFamilyBase,
        fontSize: '16px',
        fontWeight: fontWeightRegular,
        lineHeight: '1.5em',
      },
      body2: {
        fontFamily: fontFamilyBase,
        fontSize: '14px',
        fontWeight: fontWeightRegular,
        lineHeight: '1.4em',
      },
      button: {
        fontFamily: fontFamilyBase,
        fontSize: '16px',
        fontWeight: fontWeightMedium,
        textTransform: 'none',
      },
      caption: {
        fontFamily: fontFamilyBase,
        fontSize: '12px',
        fontWeight: fontWeightRegular,
        textTransform: 'none',
        lineHeight: '1.2em',
        color: palette.background.textSecondary,
      },
    });
    const result = createTypography(palette, breakpoints, typography);
    expect(result).toMatchSnapshot();
  });
});
