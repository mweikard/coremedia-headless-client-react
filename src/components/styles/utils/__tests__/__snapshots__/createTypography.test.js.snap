// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTypography() should return matching value using custom value (function) 1`] = `
Object {
  "body1": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Georgia",
    "fontSize": "16px",
    "fontWeight": 300,
    "lineHeight": "1.5em",
  },
  "body2": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Georgia",
    "fontSize": "14px",
    "fontWeight": 300,
    "lineHeight": "1.4em",
  },
  "button": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Georgia",
    "fontSize": "16px",
    "fontWeight": 700,
    "textTransform": "none",
  },
  "caption": Object {
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "Georgia",
    "fontSize": "12px",
    "fontWeight": 300,
    "lineHeight": "1.2em",
    "textTransform": "none",
  },
  "fontFamily": "Georgia",
  "fontWeightLight": 200,
  "fontWeightMedium": 700,
  "fontWeightRegular": 300,
  "htmlFontSize": 14,
  "title1": Object {
    "@media (min-width:48em)": Object {
      "fontSize": "34px",
    },
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "\\"Comic Sans\\"",
    "fontSize": "22px",
    "fontWeight": 200,
    "lineHeight": "1.6em",
    "textTransform": "uppercase",
  },
  "title2": Object {
    "@media (min-width:48em)": Object {
      "fontSize": "30px",
    },
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "\\"Comic Sans\\"",
    "fontSize": "22px",
    "fontWeight": 200,
    "lineHeight": "1.5em",
    "textTransform": "uppercase",
  },
  "title3": Object {
    "@media (min-width:48em)": Object {
      "fontSize": "26px",
    },
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "\\"Comic Sans\\"",
    "fontSize": "22px",
    "fontWeight": 200,
    "lineHeight": "1.4em",
    "textTransform": "uppercase",
  },
  "title4": Object {
    "@media (min-width:48em)": Object {
      "fontSize": "23px",
    },
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "\\"Comic Sans\\"",
    "fontSize": "22px",
    "fontWeight": 300,
    "lineHeight": "1.3em",
    "textTransform": "uppercase",
  },
  "title5": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Georgia",
    "fontSize": "20px",
    "fontWeight": 300,
    "lineHeight": "1.2em",
  },
  "title6": Object {
    "@media (min-width:48em)": Object {
      "fontSize": "18px",
    },
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Georgia",
    "fontSize": "16px",
    "fontWeight": 300,
    "lineHeight": "1.1em",
  },
}
`;

exports[`createTypography() should return matching value using custom value (object) 1`] = `
Object {
  "body1": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Verdana",
    "fontSize": "1.1428571428571428rem",
    "fontWeight": 300,
    "lineHeight": "1.3125em",
  },
  "body2": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Verdana",
    "fontSize": "1rem",
    "fontWeight": 300,
    "lineHeight": "1.2857142857142858em",
  },
  "button": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Verdana",
    "fontSize": "1.1428571428571428rem",
    "fontWeight": 700,
    "textTransform": "none",
  },
  "caption": Object {
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "Verdana",
    "fontSize": "0.8571428571428571rem",
    "fontWeight": 300,
    "lineHeight": "1.3333333333333333em",
    "textTransform": "none",
  },
  "fontFamily": "Verdana",
  "fontWeightLight": 200,
  "fontWeightMedium": 700,
  "fontWeightRegular": 300,
  "htmlFontSize": 14,
  "title1": Object {
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "Verdana",
    "fontSize": "2.4285714285714284rem",
    "fontWeight": 200,
    "lineHeight": "1.2647058823529411em",
    "textTransform": "uppercase",
  },
  "title2": Object {
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "Verdana",
    "fontSize": "2.142857142857143rem",
    "fontWeight": 200,
    "lineHeight": "1.1333333333333333em",
    "textTransform": "uppercase",
  },
  "title3": Object {
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "Verdana",
    "fontSize": "1.8571428571428572rem",
    "fontWeight": 200,
    "lineHeight": "1.1538461538461537em",
    "textTransform": "uppercase",
  },
  "title4": Object {
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "Verdana",
    "fontSize": "1.6428571428571428rem",
    "fontWeight": 300,
    "lineHeight": "1.2608695652173914em",
    "textTransform": "uppercase",
  },
  "title5": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Verdana",
    "fontSize": "1.4285714285714286rem",
    "fontWeight": 300,
    "lineHeight": "1.3em",
  },
  "title6": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Verdana",
    "fontSize": "1.2857142857142858rem",
    "fontWeight": 300,
    "lineHeight": "1.2777777777777777em",
  },
}
`;

exports[`createTypography() should return matching value using default parameter value 1`] = `
Object {
  "body1": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Helvetica, Arial, sans-serif",
    "fontSize": "1rem",
    "fontWeight": 400,
    "lineHeight": "1.3125em",
  },
  "body2": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Helvetica, Arial, sans-serif",
    "fontSize": "0.875rem",
    "fontWeight": 400,
    "lineHeight": "1.2857142857142858em",
  },
  "button": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Helvetica, Arial, sans-serif",
    "fontSize": "1rem",
    "fontWeight": 500,
    "textTransform": "none",
  },
  "caption": Object {
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "Helvetica, Arial, sans-serif",
    "fontSize": "0.75rem",
    "fontWeight": 400,
    "lineHeight": "1.3333333333333333em",
    "textTransform": "none",
  },
  "fontFamily": "Helvetica, Arial, sans-serif",
  "fontWeightLight": 300,
  "fontWeightMedium": 500,
  "fontWeightRegular": 400,
  "htmlFontSize": 16,
  "title1": Object {
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "Helvetica, Arial, sans-serif",
    "fontSize": "2.125rem",
    "fontWeight": 300,
    "lineHeight": "1.2647058823529411em",
    "textTransform": "uppercase",
  },
  "title2": Object {
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "Helvetica, Arial, sans-serif",
    "fontSize": "1.875rem",
    "fontWeight": 300,
    "lineHeight": "1.1333333333333333em",
    "textTransform": "uppercase",
  },
  "title3": Object {
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "Helvetica, Arial, sans-serif",
    "fontSize": "1.625rem",
    "fontWeight": 300,
    "lineHeight": "1.1538461538461537em",
    "textTransform": "uppercase",
  },
  "title4": Object {
    "color": "rgba(0,0,0,0.6)",
    "fontFamily": "Helvetica, Arial, sans-serif",
    "fontSize": "1.4375rem",
    "fontWeight": 400,
    "lineHeight": "1.2608695652173914em",
    "textTransform": "uppercase",
  },
  "title5": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Helvetica, Arial, sans-serif",
    "fontSize": "1.25rem",
    "fontWeight": 400,
    "lineHeight": "1.3em",
  },
  "title6": Object {
    "color": "rgba(0,0,0,0.87)",
    "fontFamily": "Helvetica, Arial, sans-serif",
    "fontSize": "1.125rem",
    "fontWeight": 400,
    "lineHeight": "1.2777777777777777em",
  },
}
`;
